pkg/apis/certmanager/v1/types_certificate.go

```diff
@@ -207,25 +207,30 @@ type CertificateSpec struct {
 
 	// Requested DNS subject alternative names.
 	// +optional
+	// +listType=set
 	DNSNames []string `json:"dnsNames,omitempty"`
 
 	// Requested IP address subject alternative names.
 	// +optional
+	// +listType=set
 	IPAddresses []string `json:"ipAddresses,omitempty"`
 
 	// Requested URI subject alternative names.
 	// +optional
+	// +listType=set
 	URIs []string `json:"uris,omitempty"`
 
 	// `otherNames` is an escape hatch for SAN that allows any type. We currently restrict the support to string like otherNames, cf RFC 5280 p 37
 	// Any UTF8 String valued otherName can be passed with by setting the keys oid: x.x.x.x and UTF8Value: somevalue for `otherName`.
 	// Most commonly this would be UPN set with oid: 1.3.6.1.4.1.311.20.2.3
 	// You should ensure that any OID passed is valid for the UTF8String type as we do not explicitly validate this.
 	// +optional
+	// +listType=atomic
```

Comment:
Let's maybe fix the `OtherName` type (make its fields non-optional) and make this a map that uses both fields as "keys" in the map.